### PR TITLE
[rfc] Comptibility header (v1)

### PR DESCRIPTION
**Pull request by alexandruradovici**
### Pull Request Overview

This pull request adds a compatibility header to the TBF. This is the first version of such a header.

The idea is to add the ability to the kernel to ignore and not load the processes that do not meet the kernel compatibility constraints defined in this header.

This defines a fixed header that contains:
- the exact kernel major version
- the minimum kernel minor version

Processes that provide a `KernelVersion` header state that the exact required kernel version and the
minimum required kernel minor version. For instance, if a process specifies kernel version 2.12, this means it requires a kernel
version that is >= 2.12 and < 3.0.

### Testing Strategy

This was tested using a Microbit v2 and libtock-c with the new elf2tab (https://github.com/tock/elf2tab/pull/30).

EDIT: Tockloader seems to send over to the board unknown TLV headers as this works with the unmodified tockloader.

### TODO or Help Wanted

1. I'm not sure where to define the running kernel version. For now this is defined as two constants in `kernel/src/lib.rs`. Another idea was to define them within the `Config` structure. The printed kernel version is taken from the git release, which, at least for the development tree, is difficult to parse
2. Improvement to the debug messages

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: doc/TockBinaryFormat.md
Comment (hudson-ayers) [inline doc/TockBinaryFormat.md:165]: These are `u8`s here but seem to be `u16`s in the tock-tbf code?

Comment (alexandruradovici) [reply doc/TockBinaryFormat.md]: Fixed, sorry for that

File: doc/TockBinaryFormat.md
Comment (bradjc) [inline doc/TockBinaryFormat.md:163]: Don't need this length.

File: kernel/src/lib.rs
Comment (bradjc) [inline kernel/src/lib.rs:94]: We need some way to remember to update this.

File: doc/TockBinaryFormat.md
Comment (phil-levis) [inline doc/TockBinaryFormat.md:384]: I am wary of this kind of specification. For example, it means that we can't change a system call API for a particular device number without a major number increment. Defining "compatible" seems difficult. 

Instead, I'd suggest that we add this to the end of the main header. It should either be present in every app, or not. It should include only a major version number, which relates to the ABI with the kernel. Anything further than that (compatibility) can be determined by the app itself, e.g., by probing if devices are installed.

Comment (bradjc) [reply doc/TockBinaryFormat.md:384]: The intent is for this (i.e. TLV id 7) to only specify kernel ABI. We wanted major and minor, because Tock 2.1 might include a new allow call that Tock 2.0 does not. We do need to make this clear in the docs.

We also thought that a TLV was an appropriate option for two reasons:
1. We're not entirely sure this is the long term best way to handle this sort of compatibility checking. Using a TLV makes it easy to deprecate this in the future and possibly replace it with something else.
2. Some boards/deployments might not want to include any compatibility information (and save the bytes in the header). They can easily omit the TLV.

Comment (alexandruradovici) [reply doc/TockBinaryFormat.md:384]: I detailed this in the documentation.

File: kernel/src/process_standard.rs
Comment (bradjc) [inline kernel/src/process_standard.rs:1425]: I think we should return an error here. The caller asked to enforce kernel version, and the check failed.

File: kernel/src/process_standard.rs
Comment (bradjc) [inline kernel/src/process_standard.rs:1411]: Do we need to separate these cases?

File: kernel/src/process_standard.rs
Comment (bradjc) [inline kernel/src/process_standard.rs:1413]: I think this should error if `enforce_kernel_version` is true.

File: kernel/src/process_standard.rs
Comment (alexandruradovici) [reply kernel/src/process_standard.rs:1411]: If we want to display different errors, but I suppose we could merge this into a single if.

File: kernel/src/process_standard.rs
Comment (alexandruradovici) [reply kernel/src/process_standard.rs:1425]: My understanding is that if this returns an error, `load_processes_advanced` will stop loading processes, not just skip this process.

```rust
  ProcessStandard::create(
                    kernel,
                    chip,
                    entry_flash,
                    header_length as usize,
                    version,
                    remaining_memory,
                    fault_policy,
                    enforce_kernel_version,
                    i,
            )?
```

Comment (bradjc) [reply kernel/src/process_standard.rs:1421]: That's correct. But without an error this will silently fail, and it's hard to differentiate an app that just doesn't do anything from one that was never actually loaded. Presumably no user who wants the kernel version compatibility checked wants to just ignore those errors.

Comment (alexandruradovici) [reply kernel/src/process_standard.rs:1421]: Agree, but on the other hand this would make "some apps" (all apps after the one that cannot be loaded) silently fail.

What about always showing this warnings if enforce kernel version is true?

File: kernel/src/process_standard.rs
Comment (ppannuto) [inline kernel/src/process_standard.rs:1413]:
```suggestion
                // If enforcing the kernel version is requested, and the `KernelVersion` header is not present,
```

File: doc/TockBinaryFormat.md
Comment (alistair23) [inline doc/TockBinaryFormat.md:373]: Don't remove this

File: kernel/src/lib.rs
Comment (alistair23) [reply kernel/src/lib.rs:94]: Could we auto generate/update this from git tags?

File: kernel/src/process_standard.rs
Comment (bradjc) [reply kernel/src/process_standard.rs:1421]: What is the use case for knowing that you have an incompatible app but continuing to run the others? I suspect that the most likely case is the user only has one app (or if there are multiple they are all incompatible) and they are going to fix it immediately.

File: doc/TockBinaryFormat.md
Comment (alexandruradovici) [reply doc/TockBinaryFormat.md:432]: Do you mean the empty new line?

File: kernel/src/process_standard.rs
Comment (alexandruradovici) [reply kernel/src/process_standard.rs:1422]: The user might have loaded one of the apps from a tab using tockloader. The issue is that it will load all apps up to the one that does not meet the requirements or has no header. In case we do not want to load any app, maybe it would be a good idea to panic as it is not deterministic which apps will run.

Comment (bradjc) [reply kernel/src/process_standard.rs:1422]: Most boards do not panic, but do print a warning/error when load_processes fails. I've heard anecdotally that the checks for an app that does not meet ram/flash position requirements which cause load_processes to fail have worked well.

File: kernel/src/lib.rs
Comment (bradjc) [reply kernel/src/lib.rs:94]: Generate doesn't work if someone downloads the zip file and there is no git repo. But something that at least checked
somehow would be nice.

File: doc/TockBinaryFormat.md
Comment (hudson-ayers) [reply doc/TockBinaryFormat.md:432]: yes he does

File: kernel/src/process_standard.rs
Comment (alexandruradovici) [reply kernel/src/process_standard.rs]: I got it now :)  

I just saw that most boards report a debug message upon an error.

File: kernel/src/process_standard.rs
Comment (bradjc) [inline kernel/src/process_standard.rs:1343]:
```suggestion
        require_kernel_version: bool,
```

File: kernel/src/process_standard.rs
Comment (bradjc) [inline kernel/src/process_standard.rs:1415]:
```suggestion
            if require_kernel_version {
```

I think the current logic makes sense: if a process tells us it requires a newer kernel version we always error in that case. If the process doesn't tell us what it requires we assume the process is fine, unless explicitly told not to. 

We may someday want to add a "force allow incompatible apps" or something, but we can make that change if the need ever arises.

File: kernel/src/process_utilities.rs
Comment (bradjc) [inline kernel/src/process_utilities.rs:53]:
```suggestion
    /// A process requires a newer version of the kernel or did not specify
    /// a required version. Processes can include the KernelVersion TBF header stating
    /// their compatible kernel version (^major.minor).
    ///
    /// Boards may not require processes to include the KernelVersion TBF header, and
    /// the kernel supports ignoring a missing KernelVersion TBF header. In that case,
    /// this error will not be returned for a process missing a KernelVersion TBF
    /// header.
    ///
    /// `version` is the `(major, minor)` kernel version the process indicates it
    /// requires. If `version` is `None` then the process did not include the
    /// KernelVersion TBF header.
```

File: kernel/src/process_utilities.rs
Comment (bradjc) [inline kernel/src/process_utilities.rs:277]:
```suggestion
///  - `require_kernel_version`: prevent loading processes that do not provide a `KernelVersion`
```

File: libraries/tock-tbf/src/types.rs
Comment (bradjc) [inline libraries/tock-tbf/src/types.rs:466]:
```suggestion
    /// Get the minimum compatible kernel version this process requires.
    /// Returns `None` if the kernel compatibility header is not included.
    pub fn get_kernel_version(&self) -> Option<(u16, u16)> {
```

File: kernel/src/process_utilities.rs
Comment (bradjc) [inline kernel/src/process_utilities.rs:155]:
```suggestion
    require_kernel_version: bool,
```

File: kernel/src/process_utilities.rs
Comment (bradjc) [inline kernel/src/process_utilities.rs:242]:
```suggestion
                    require_kernel_version,
```

File: kernel/src/process_utilities.rs
Comment (bradjc) [inline kernel/src/process_utilities.rs:112]:
```suggestion
                    "Process is incompatible with the kernel. Running: {}.{}, Requested: {}.{}",
```

File: kernel/src/process_utilities.rs
Comment (bradjc) [inline kernel/src/process_utilities.rs:118]:
```suggestion
                None => write!(f, "Process did not provide a TBF kernel version header"),
```

File: kernel/src/process_standard.rs
Comment (bradjc) [inline kernel/src/process_standard.rs:1390]: I don't understand this change. Is this related?

File: doc/TockBinaryFormat.md
Comment (bradjc) [inline doc/TockBinaryFormat.md:438]:
```suggestion
It defines the following two items:
```

File: doc/TockBinaryFormat.md
Comment (bradjc) [inline doc/TockBinaryFormat.md:447]:
```suggestion
A kernel major and minor version guarantees the ABI for exchanging 
```

File: kernel/src/process_standard.rs
Comment (alexandruradovici) [reply kernel/src/process_standard.rs:1390]: Its not related, but this would display the process name instead of Some(process_name). I suggest the same change at line 1452.

File: kernel/src/process_utilities.rs
Comment (hudson-ayers) [inline kernel/src/process_utilities.rs:154]:
```suggestion
#[inline(always)]
pub fn load_processes_advanced<C: Chip>(
```

Given that load_processes has the annotation we should it include it here as well.

File: kernel/src/process_utilities.rs
Comment (hudson-ayers) [reply kernel/src/process_utilities.rs:283]: This was marked as resolved but does not look to be changed?

File: kernel/src/process_utilities.rs
Comment (alexandruradovici) [reply kernel/src/process_utilities.rs:155]: I added the inline option to load_processes as it is a wrapper for load_processes_advanced. The original load_processes did not have this option.

**Review by bradjc — APPROVED** [review]
We need to merge at least the elf2tab changes and release elf2tab before we merge this.

**Comment by bradjc** [comment]
Or we could set required to `false` for now, and change it back closer to the actual 2.0 release.

File: kernel/src/process_utilities.rs
Comment (alexandruradovici) [reply kernel/src/process_utilities.rs:155]: @hudson-ayers it's changed now, sorry I forgot to rename the variable in `process_utilities.rs`.

**Comment by alexandruradovici** [comment]
> Or we could set required to `false` for now, and change it back closer to the actual 2.0 release.

As long is we change it back before we release 2.0 this should be fine.

**Comment by phil-levis** [comment]
Let's do this right -- @alexandruradovici can you rebase https://github.com/tock/elf2tab/pull/30 so we can review and merge? It seems like everyone is on-board and it all looks good, let's just do it.

**Comment by alexandruradovici** [comment]
I rebased https://github.com/tock/elf2tab/pull/30, I would be great if @alistair23 can take a look at it the merge was with the permissions header.

**Review by lschuermann — APPROVED** [review]

**Pull request merged**
